src/sync/spsc_bufferqueue.rs

```diff
@@ -0,0 +1,400 @@
+    tail_cache: AtomicUsize,
+    prod_alive:AtomicBool, //seems weird, but consumer will read this
+
+    _dummy_2: CachePadded<u64>,
+    // data for the producer
+    tail: AtomicUsize,
+    head_cache: AtomicUsize,
+    cons_alive: AtomicBool, //seems weird, but producer will read this
+}
+
+unsafe impl<T: Send> Send for SpscBufferQueue<T> {}
+
+impl<T: Send> SpscBufferQueue<T> {
+    pub fn new(size: usize) -> (BufferProducer<T>, BufferConsumer<T>) {
+        let q = SpscBufferQueue {
+            data_block: unsafe{ allocate(size) },
+            size: cmp::min(size, (isize::max_value() - 1) as usize) + 1,
+            _marker: PhantomData,
+
+            _dummy_1: CachePadded::zeroed(),
+            head: AtomicUsize::new(0),
+            tail_cache: AtomicUsize::new(0),
+            prod_alive: AtomicBool::new(true),
+
+            _dummy_2: CachePadded::zeroed(),
+            tail: AtomicUsize::new(0),
+            head_cache: AtomicUsize::new(0),
+            cons_alive: AtomicBool::new(true),
+        };
+        let qarc = Arc::new(q);
+        let rtuple = (BufferProducer::new(qarc.clone()),
+                      BufferConsumer::new(qarc));
+        fence(SeqCst);
+        rtuple
+    }
+
+    /// Tries constructing the element and inserts into the queue
+    ///
+    /// Returns the closure if there isn't space
+    #[inline(always)]
+    fn try_construct<F>(&self, ctor: F) -> Result<(), F> where F: FnOnce() -> T {
+        let ctail = self.tail.load(Relaxed);
+        let mut next_tail = ctail + 1;
+        next_tail = if next_tail == self.size  { 0 } else { next_tail };
+        if next_tail == self.head_cache.load(Relaxed) {
+            let cur_head = self.head.load(Acquire);
+            self.head_cache.store(cur_head, Relaxed);
+            if next_tail == cur_head {
+                return Err(ctor);
+            }
+        }
+        unsafe {
+            let data_pos = self.data_block.offset(ctail as isize);
+            ptr::write(data_pos, ctor());
+        }
+        self.tail.store(next_tail, Release);
+        Ok(())
+    }
+
+    pub fn try_pop(&self) -> Option<T> {
+        let chead = self.head.load(Relaxed);
+        if chead == self.tail_cache.load(Relaxed) {
+            let cur_tail = self.tail.load(Acquire);
+            self.tail_cache.store(cur_tail, Relaxed);
+            if chead == cur_tail {
+                return None;
+            }
+        }
+
+        let mut next_head = chead + 1;
+        next_head = if next_head == self.size  { 0 } else { next_head };
+        unsafe {
+            let data_pos = self.data_block.offset(chead as isize);
+            let rval = Some(ptr::read(data_pos));
+            self.head.store(next_head, Release);
+            rval
+        }
+    }
+
+    pub fn capacity(&self) -> usize {
+        self.size - 1 //extra space added in ctor as buffer for head/tail
+    }
+}
+
+
+impl<T: Send> Drop for SpscBufferQueue<T> {
+    fn drop(&mut self) {
+        fence(SeqCst);
```

Comment:
It was originally intended to ensure that the destructor only runs once all previous memory accesses have happened. However, since that's already guarded by the Arc, it's kinda pointless now.